File: 2022-04/javascript/707_Design_Linked_List.js
class MyListNode {
    constructor(val) {
        this.val = val
        this.next = null
    }
}

class MyLinkedList {
    constructor() {
        this.head = null
        this.size = 0
    }

    addAtIndex(index, val) {
        if (index < 0 || index > this.size) return
        const newNode = new MyListNode(val)
        this.size++
        if (index === 0) {
            newNode.next = this.head
            this.head = newNode
        } else {
            let prev = this.head
            while (index > 1) {
                prev = prev.next
                index--
            }
            newNode.next = prev.next
            prev.next = newNode
        }
    }

    get(index) {
        if (index < 0 || index >= this.size) return -1
        let target = this.head
        while (index > 0) {
            target = target.next
            index--
        }
        return target.val
    }

    deleteAtIndex(index) {
        if (index >= 0 && index < this.size) {
            this.size--
            if (index === 0) {
                this.head = this.head.next
            } else {
                let prev = this.head
                while (index > 1) {
                    prev = prev.next
                    index--
                }
                prev.next = prev.next.next
            }
        }
    }

    addAtHead(val) {
        this.addAtIndex(0, val)
    }

    addAtTail(val) {
        this.addAtIndex(this.size, val)
    }
}

File: 2022-04/python/707_Design_Linked_List.py
class MyListNode:

    def __init__(self, val: int):
        self.val = val
        self.next = None


class MyLinkedList:

    def __init__(self):
        self.head = None
        self.size = 0

    def addAtIndex(self, index: int, val: int) -> None:
        if 0 <= index <= self.size:
            new_node = MyListNode(val)
            self.size += 1
            if index == 0:
                new_node.next = self.head
                self.head = new_node
            else:
                prev = self.head
                while index > 1:
                    index -= 1
                    prev = prev.next
                new_node.next = prev.next
                prev.next = new_node

    def get(self, index: int) -> int:
        if 0 <= index < self.size:
            target = self.head
            while index > 0:
                index -= 1
                target = target.next
            return target.val
        return -1

    def deleteAtIndex(self, index: int) -> None:
        if 0 <= index < self.size:
            self.size -= 1
            if index == 0:
                self.head = self.head.next
            else:
                prev = self.head
                while index > 1:
                    index -= 1
                    prev = prev.next
                prev.next = prev.next.next

    def addAtHead(self, val: int) -> None:
        self.addAtIndex(0, val)

    def addAtTail(self, val: int) -> None:
        self.addAtIndex(self.size, val)